From cd51042a92b346cea7ced5d1a42a51c1eb94c275 Mon Sep 17 00:00:00 2001
From: Icenowy Zheng <icenowy@aosc.io>
Date: Thu, 16 Aug 2018 22:03:27 +0800
Subject: [PATCH 036/131] drm: sun4i: Get the HDMI PHY clock exclusive in
 sun8i-dw-hdmi

The HDMI PHY clock is also sensitive to unattended changes.

Get it exclusive just after setting up it, and drop the exclusivity when
disabling the PHY. The balence of exclusivity is manually maintained due
to the dw-hdmi driver being not clearly documented.

Signed-off-by: Icenowy Zheng <icenowy@aosc.io>
---
 drivers/gpu/drm/sun4i/sun8i_dw_hdmi.h  |  1 +
 drivers/gpu/drm/sun4i/sun8i_hdmi_phy.c | 15 ++++++++++++++-
 2 files changed, 15 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/sun4i/sun8i_dw_hdmi.h b/drivers/gpu/drm/sun4i/sun8i_dw_hdmi.h
index 556a32260207..9b0c1ec7fe22 100644
--- a/drivers/gpu/drm/sun4i/sun8i_dw_hdmi.h
+++ b/drivers/gpu/drm/sun4i/sun8i_dw_hdmi.h
@@ -168,6 +168,7 @@ struct sun8i_hdmi_phy {
 	struct clk			*clk_phy;
 	struct clk			*clk_pll0;
 	struct clk			*clk_pll1;
+	bool				clk_phy_exclusive;
 	unsigned int			rcal;
 	struct regmap			*regs;
 	struct reset_control		*rst_phy;
diff --git a/drivers/gpu/drm/sun4i/sun8i_hdmi_phy.c b/drivers/gpu/drm/sun4i/sun8i_hdmi_phy.c
index 66ea3a902e36..e3327567b25b 100644
--- a/drivers/gpu/drm/sun4i/sun8i_hdmi_phy.c
+++ b/drivers/gpu/drm/sun4i/sun8i_hdmi_phy.c
@@ -354,8 +354,14 @@ static int sun8i_hdmi_phy_config(struct dw_hdmi *hdmi, void *data,
 	regmap_update_bits(phy->regs, SUN8I_HDMI_PHY_DBG_CTRL_REG,
 			   SUN8I_HDMI_PHY_DBG_CTRL_POL_MASK, val);
 
-	if (phy->variant->has_phy_clk)
+	if (phy->variant->has_phy_clk) {
+		if (!phy->clk_phy_exclusive) {
+			clk_rate_exclusive_get(phy->clk_phy);
+			phy->clk_phy_exclusive = true;
+		}
+
 		clk_set_rate(phy->clk_phy, mode->crtc_clock * 1000);
+	}
 
 	return phy->variant->phy_config(hdmi, phy, mode->crtc_clock * 1000);
 };
@@ -363,6 +369,7 @@ static int sun8i_hdmi_phy_config(struct dw_hdmi *hdmi, void *data,
 static void sun8i_hdmi_phy_disable_a83t(struct dw_hdmi *hdmi,
 					struct sun8i_hdmi_phy *phy)
 {
+	/* A83T variant doesn't have PHY clock */
 	dw_hdmi_phy_gen2_txpwron(hdmi, 0);
 	dw_hdmi_phy_gen2_pddq(hdmi, 1);
 
@@ -373,6 +380,12 @@ static void sun8i_hdmi_phy_disable_a83t(struct dw_hdmi *hdmi,
 static void sun8i_hdmi_phy_disable_h3(struct dw_hdmi *hdmi,
 				      struct sun8i_hdmi_phy *phy)
 {
+	/* H3 variant has PHY clock */
+	if (phy->clk_phy_exclusive) {
+		clk_rate_exclusive_put(phy->clk_phy);
+		phy->clk_phy_exclusive = false;
+	}
+
 	regmap_write(phy->regs, SUN8I_HDMI_PHY_ANA_CFG1_REG,
 		     SUN8I_HDMI_PHY_ANA_CFG1_LDOEN |
 		     SUN8I_HDMI_PHY_ANA_CFG1_ENVBS |
-- 
2.18.1

